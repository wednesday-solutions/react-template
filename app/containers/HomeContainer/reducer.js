/*
 *
 * HomeContainer reducer
 *
 */
import { produce } from 'immer';
import { createActions } from 'reduxsauce';
import get from 'lodash/get';

export const { Types: homeContainerTypes, Creators: homeContainerCreators } = createActions({
  requestGetGithubRepos: ['repoName'],
  successGetGithubRepos: ['data'],
  failureGetGithubRepos: ['error'],
  clearGithubRepos: {}
});
export const initialState = { repoName: null, reposData: {}, reposError: null, loading: null };

export const homeContainerReducer = (state = initialState, action) =>
  produce(state, (draft) => {
    switch (action.type) {
      case homeContainerTypes.REQUEST_GET_GITHUB_REPOS:
        draft.repoName = action.repoName;
        draft.loading = true;
        break;
      case homeContainerTypes.CLEAR_GITHUB_REPOS:
        draft.repoName = null;
        draft.reposError = null;
        draft.reposData = {};
        draft.loading = null;
        break;
      case homeContainerTypes.SUCCESS_GET_GITHUB_REPOS:
        draft.reposData = action.data;
        draft.reposError = null;
        draft.loading = false;
        break;
      case homeContainerTypes.FAILURE_GET_GITHUB_REPOS:
        draft.reposError = get(action.error, 'message', 'something_went_wrong');
        draft.reposData = null;
        draft.loading = false;
        break;
    }
  });

export default homeContainerReducer;
