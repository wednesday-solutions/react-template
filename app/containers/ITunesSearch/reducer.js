import produce from 'immer';
import { SEARCH_TRACKS, SEARCH_TRACKS_SUCCESS, SEARCH_TRACKS_ERROR, CLEAR_TRACKS } from './constants';

export const initialState = {
  tracks: [],
  loading: false,
  error: null,
  query: ''
};

/**
 * iTunes search reducer
 * @param {object} state - The current state
 * @param {object} action - The action to handle
 * @returns {object} The new state
 */
const itunesSearchReducer = (state = initialState, action) =>
  produce(state, (draft) => {
    switch (action.type) {
      case SEARCH_TRACKS:
        draft.loading = true;
        draft.error = null;
        draft.query = action.query;
        break;

      case SEARCH_TRACKS_SUCCESS:
        draft.tracks = action.tracks;
        draft.loading = false;
        break;

      case SEARCH_TRACKS_ERROR:
        draft.error = action.error;
        draft.loading = false;
        break;

      case CLEAR_TRACKS:
        draft.tracks = [];
        draft.error = null;
        draft.query = '';
        break;
    }
  });

export default itunesSearchReducer;
