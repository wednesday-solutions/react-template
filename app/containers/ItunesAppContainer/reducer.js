/*
 *
 * ItunesAppContainer reducer
 *
 */
import produce from 'immer';
import { createActions } from 'reduxsauce';
import { fromJS } from 'immutable';

export const initialState = fromJS({ songsData: [], showLoader: false });

export const { Types: itunesAppContainerTypes, Creators: itunesAppContainerCreators } = createActions({
  setSongs: ['songs'],
  requestSearchSong: ['artistName'],
  setLoader: ['bool']
});

/* eslint-disable default-case, no-param-reassign */
export const itunesAppContainerReducer = (state = initialState, action) =>
  produce(state, draft => {
    switch (action.type) {
      case itunesAppContainerTypes.SET_SONGS:
        return state.set('songsData', action.songs);
      case itunesAppContainerTypes.SET_LOADER:
        return state.set('showLoader', action.bool);
      default:
        return state;
    }
  });

export default itunesAppContainerReducer;
