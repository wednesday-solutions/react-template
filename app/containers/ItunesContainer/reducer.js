/*
 *
 * HomeContainer reducer
 *
 */
import produce from 'immer';
import { createActions } from 'reduxsauce';
import get from 'lodash/get';

export const { Types: itunesContainerTypes, Creators: itunesContainerCreators } = createActions({
  requestGetItunesSongs: ['songName'],
  successGetItunesSongs: ['data'],
  failureGetItunesSongs: ['error'],
  clearItunesSongs: []
});
export const initialState = { songName: null, songsData: [], songsError: null };

/* eslint-disable default-case, no-param-reassign */
export const itunesContainerReducer = (state = initialState, action) =>
  produce(state, draft => {
    switch (action.type) {
      case itunesContainerTypes.REQUEST_GET_ITUNES_SONGS:
        draft.songName = action.songName;
        break;
      case itunesContainerTypes.CLEAR_ITUNES_SONGS:
        return initialState;
      case itunesContainerTypes.SUCCESS_GET_ITUNES_SONGS:
        draft.songsData = action.data;
        break;
      case itunesContainerTypes.FAILURE_GET_ITUNES_SONGS:
        draft.songsError = get(action.error, 'message', 'something_went_wrong');
        break;
    }
  });

export default itunesContainerReducer;
