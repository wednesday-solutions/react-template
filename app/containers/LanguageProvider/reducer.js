/*
 *
 * LanguageProvider reducer
 *
 */
import { createActions } from 'reduxsauce';
import { produce } from 'immer';
import { DEFAULT_LOCALE } from '@app/i18n';

export const { Types: languageProviderTypes, Creators: languageProviderActions } = createActions({
  changeLocale: ['locale']
});

export const initialState = {
  locale: DEFAULT_LOCALE
};

export const languageProviderReducer = (state = initialState, action) =>
  produce(state, (draft) => {
    // the number of languages should and will increase.
    // eslint-disable-next-line sonarjs/no-small-switch
    switch (action.type) {
      case languageProviderTypes.CHANGE_LOCALE:
        draft.locale = action.locale;
        break;
    }
  });

export default languageProviderReducer;
