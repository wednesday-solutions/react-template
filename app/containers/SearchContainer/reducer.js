/*
 *
 * SearchContainer reducer
 *
 */
import produce from 'immer'
import { createActions } from 'reduxsauce'

export const initialState = {}

export const { Types: searchContainerTypes, Creators: searchContainerCreators } = createActions({
  requestGetArtistSongs: ['songName'],
  successGetArtistSongs: ['data'],
  failureGetArtistSongs: ['error'],
  clearArtistSongs: []
})

/* eslint-disable default-case, no-param-reassign */
export const searchContainerReducer = (state = initialState, action) =>
  produce(state, (/* draft */) => {
    switch (action.type) {
      case searchContainerTypes.DEFAULT_ACTION:
        return {...state, somePayload: action.somePayload}
      default:
        return state
    }
  })

export default searchContainerReducer
