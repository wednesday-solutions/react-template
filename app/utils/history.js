import { createBrowserHistory } from 'history';
import { isUAT } from './index';

export function getBaseName() {
  if (isUAT()) {
    return `/${process.env.BRANCH_NAME}`;
  }

  return '';
}

const history = createBrowserHistory({ basename: getBaseName() });
export default history;
