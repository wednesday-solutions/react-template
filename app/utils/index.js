import find from 'lodash/find';
import get from 'lodash/get';
import { i18n } from '@lingui/core';
import routeConstants from './routeConstants';

/**
 * Get details of the current route from the route config.
 * If location.pathname doesn't match any entry in the routeConstants it will return null
 * @author mac
 * @date 2020-05-07
 * @param {any} location
 * @returns {any}
 */
export const getCurrentRouteDetails = (location) => {
  if (!get(location, 'pathname')) {
    return null;
  }
  const route = find(
    Object.keys(routeConstants),
    (key) => routeConstants[key].route === location.pathname || `${routeConstants[key].route}/` === location.pathname
  );
  if (route) {
    return routeConstants[route];
  }
  return null;
};
export const mapKeysDeep = (obj, fn) => {
  if (Array.isArray(obj)) {
    return obj.map((val) => mapKeysDeep(val, fn));
  }
  if (typeof obj === 'object') {
    return Object.keys(obj).reduce((acc, current) => {
      const key = fn(current);
      const val = obj[current];
      acc[key] = val !== null && typeof val === 'object' ? mapKeysDeep(val, fn) : val;
      return acc;
    }, {});
  }
  return obj;
};

export const isLocal = () => {
  try {
    const local = JSON.parse(process.env.IS_LOCAL);
    return typeof local === 'boolean' && local;
  } catch {
    // continue regardless of error
  }
  return false;
};

export function isUAT() {
  return process.env.ENVIRONMENT_NAME === 'development' && process.env.NODE_ENV === 'production';
}

export function isProd() {
  return process.env.ENVIRONMENT_NAME === 'production' && process.env.NODE_ENV === 'production';
}

export const translate = (id, values = {}) => i18n._({ id, values });
