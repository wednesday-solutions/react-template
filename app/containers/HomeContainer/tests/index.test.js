/**
 *
 * Tests for HomeContainer
 *
 */

import React from 'react';
import { Router } from 'react-router';
import { useHistory } from 'react-router-dom';
import { fireEvent } from '@testing-library/dom';
import { timeout, renderProvider } from '@utils/testUtils';
import { HomeContainerTest as HomeContainer, mapDispatchToProps } from '../index';
import { homeContainerTypes } from '../reducer';
import { createBrowserHistory } from 'history';
import { translate } from '@app/utils/index';

describe('<HomeContainer /> tests', () => {
  let submitSpy;

  beforeEach(() => {
    submitSpy = jest.fn();
  });
  it('should render and match the snapshot', () => {
    const { baseElement } = renderProvider(<HomeContainer dispatchGithubRepos={submitSpy} />);
    expect(baseElement).toMatchSnapshot();
  });

  it('should call dispatchClearGithubRepos on empty change', async () => {
    const getGithubReposSpy = jest.fn();
    const clearGithubReposSpy = jest.fn();
    const { getByTestId } = renderProvider(
      <HomeContainer dispatchClearGithubRepos={clearGithubReposSpy} dispatchGithubRepos={getGithubReposSpy} />
    );
    fireEvent.change(getByTestId('search-bar'), {
      target: { value: 'a' }
    });
    await timeout(500);
    expect(getGithubReposSpy).toBeCalled();
    fireEvent.change(getByTestId('search-bar'), {
      target: { value: '' }
    });
    await timeout(500);
    expect(clearGithubReposSpy).toBeCalled();
  });

  it('should call dispatchGithubRepos on change and after enter', async () => {
    const repoName = 'react-template';
    const { getByTestId } = renderProvider(<HomeContainer dispatchGithubRepos={submitSpy} />);
    const searchBar = getByTestId('search-bar');
    fireEvent.change(searchBar, {
      target: { value: repoName }
    });
    await timeout(500);
    expect(submitSpy).toBeCalledWith(repoName);

    fireEvent.keyDown(searchBar, {
      key: 'Enter',
      code: 13,
      charCode: 13
    });
    expect(submitSpy).toBeCalledWith(repoName);
  });

  it('should call dispatchGithubRepos on clicking the search icon', async () => {
    const repoName = 'react-template';
    const { getByTestId } = renderProvider(<HomeContainer dispatchGithubRepos={submitSpy} repoName={repoName} />);
    fireEvent.click(getByTestId('search-icon'));

    await timeout(500);
    expect(submitSpy).toBeCalledWith(repoName);
  });

  it('should  dispatchGithubRepos on update on mount if repoName is already persisted', async () => {
    const repoName = 'react-template';
    renderProvider(<HomeContainer repoName={repoName} reposData={null} dispatchGithubRepos={submitSpy} />);

    await timeout(500);
    expect(submitSpy).toBeCalledWith(repoName);
  });

  it('should validate mapDispatchToProps actions', async () => {
    const dispatchReposSearchSpy = jest.fn();
    const repoName = 'react-template';
    const actions = {
      dispatchGithubRepos: { repoName, type: homeContainerTypes.REQUEST_GET_GITHUB_REPOS },
      dispatchClearGithubRepos: { type: homeContainerTypes.CLEAR_GITHUB_REPOS }
    };

    const props = mapDispatchToProps(dispatchReposSearchSpy);
    props.dispatchGithubRepos(repoName);
    expect(dispatchReposSearchSpy).toHaveBeenCalledWith(actions.dispatchGithubRepos);

    await timeout(500);
    props.dispatchClearGithubRepos();
    expect(dispatchReposSearchSpy).toHaveBeenCalledWith(actions.dispatchClearGithubRepos);
  });

  it('should render default error message when search goes wrong', () => {
    const defaultError = translate('something_went_wrong');
    const { getByTestId } = renderProvider(<HomeContainer reposError={defaultError} />);
    expect(getByTestId('error-message')).toBeInTheDocument();
    expect(getByTestId('error-message').textContent).toBe(defaultError);
  });

  it('should render the default message when searchBox is empty and reposError is null', () => {
    const defaultMessage = translate('repo_search_default');
    const { getByTestId } = renderProvider(<HomeContainer />);
    expect(getByTestId('default-message')).toBeInTheDocument();
    expect(getByTestId('default-message').textContent).toBe(defaultMessage);
  });

  it('should render the data when loading becomes false', () => {
    const reposData = { items: [{ repoOne: 'react-template' }] };
    const { getByTestId } = renderProvider(<HomeContainer reposData={reposData} dispatchGithubRepos={submitSpy} />);
    expect(getByTestId('for')).toBeInTheDocument();
  });

  it('should render exact number of RepoCards as per totalCount in result', () => {
    const totalCount = 3;
    const reposData = {
      totalCount,
      items: [
        {
          name: 'react-tempalte',
          fullName: 'wednesday-solutions/react-template',
          stargazersCount: 200
        },
        {
          name: 'react',
          fullName: 'wednesday-solutions/react',
          stargazersCount: 100
        },
        {
          name: 'react-tempalte2',
          fullName: 'wednesday-solutions/react-template2',
          stargazersCount: 300
        }
      ]
    };
    const { getAllByTestId } = renderProvider(<HomeContainer reposData={reposData} dispatchGithubRepos={submitSpy} />);
    expect(getAllByTestId('repo-card').length).toBe(totalCount);
  });

  it('should redirect to /stories when clicked on Clickable component', async () => {
    const history = createBrowserHistory();
    const { getByTestId } = renderProvider(
      <Router history={history}>
        <HomeContainer />
      </Router>
    );
    const h = useHistory();
    const historySpy = jest.spyOn(h, 'push');
    fireEvent.click(getByTestId('redirect'));
    await timeout(500);
    expect(historySpy).toHaveBeenCalledWith('/stories');
  });

  it('should render Skeleton Comp when "loading" is true', async () => {
    const repoName = 'some repo';
    const { getByTestId, getAllByTestId } = renderProvider(
      <HomeContainer dispatchGithubRepos={submitSpy} repoName={repoName} />
    );
    fireEvent.change(getByTestId('search-bar'), { target: { value: repoName } });
    await timeout(500);
    expect(getAllByTestId('skeleton').length).toBe(3);
  });
});
