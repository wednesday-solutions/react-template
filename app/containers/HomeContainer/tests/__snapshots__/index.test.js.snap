// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<HomeContainer /> tests should render and match the snapshot 1`] = `
<body>
  <div>
    <div
      class="css-6azwsi e666d1v3"
    >
      <div
        class="css-1cnh6jp e666d1v2"
      >
        <p
          class="e666d1v1 css-gnuwwn egtqi0h0"
          data-testid="redirect"
        >
          Go to Storybook
        </p>
      </div>
      <div
        class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root e666d1v4 css-1azwflu-MuiPaper-root-MuiCard-root"
        maxwidth="500"
        title="Repository Search"
      >
        <p
          class="css-darbe1 egtqi0h0"
          data-testid="t"
        >
          Get details of repositories
        </p>
        <div
          class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-adornedEnd e666d1v0 css-16qisbh-MuiInputBase-root-MuiOutlinedInput-root"
        >
          <input
            class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputAdornedEnd css-nxo287-MuiInputBase-input-MuiOutlinedInput-input"
            data-testid="search-bar"
            type="text"
            value=""
          />
          <div
            class="MuiInputAdornment-root MuiInputAdornment-positionEnd css-1laqsz7-MuiInputAdornment-root"
          >
            <button
              aria-label="search repos"
              class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-78trlr-MuiButtonBase-root-MuiIconButton-root"
              data-testid="search-icon"
              tabindex="0"
              type="button"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                data-testid="SearchIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M15.5 14h-.79l-.28-.27C15.41 12.59 16 11.11 16 9.5 16 5.91 13.09 3 9.5 3S3 5.91 3 9.5 5.91 16 9.5 16c1.61 0 3.09-.59 4.23-1.57l.27.28v.79l5 4.99L20.49 19l-4.99-5zm-6 0C7.01 14 5 11.99 5 9.5S7.01 5 9.5 5 14 7.01 14 9.5 11.99 14 9.5 14z"
                />
              </svg>
              <span
                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              />
            </button>
          </div>
          <fieldset
            aria-hidden="true"
            class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
          >
            <legend
              class="css-ihdtdm"
            >
              <span
                class="notranslate"
              >
                â€‹
              </span>
            </legend>
          </fieldset>
        </div>
      </div>
      <div
        class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root e666d1v4 css-ckxzvw-MuiPaper-root-MuiCard-root"
        color="grey"
        title="Repository List"
      >
        <p
          class="css-lh9t18 egtqi0h0"
          data-testid="default-message"
        >
          Search for a repository by entering it's name in the search box
        </p>
      </div>
    </div>
  </div>
</body>
`;
